File: src/module/daily/daily_protocol.erl
-module(daily_protocol).
-export([read/2, write/2]).
-include("count.hrl").
-include("daily.hrl").


-spec read(Protocol :: non_neg_integer(), Binary :: binary()) -> {ok, [integer() | binary() | list()]} | {error, Protocol :: non_neg_integer(), Binary :: binary()}.
read(12301, <<>>) ->
    {ok, []};

read(12302, <<>>) ->
    {ok, []};

read(12303, <<DailyId:32>>) ->
    {ok, DailyId};

read(12304, <<StageId:32>>) ->
    {ok, StageId};

read(Protocol, Binary) ->
    {error, Protocol, Binary}.


-spec write(Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()) -> {ok, binary()} | {error, Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()}.
write(12301, List) ->
    ListBinary = protocol:write_list(fun(#count{type = Type, today_number = TodayNumber}) -> <<Type:32, TodayNumber:32>> end, List),
    {ok, protocol:pack(12301, <<ListBinary/binary>>)};

write(12302, [List, #daily_active{stage_id = StageId, score = Score}]) ->
    ListBinary = protocol:write_list(fun(#daily{daily_id = DailyId, is_award = IsAward}) -> <<DailyId:32, IsAward:8>> end, List),
    {ok, protocol:pack(12302, <<ListBinary/binary, StageId:32, Score:32>>)};

write(12303, Result) ->
    {ok, protocol:pack(12303, <<(protocol:text(Result))/binary>>)};

write(12304, Result) ->
    {ok, protocol:pack(12304, <<(protocol:text(Result))/binary>>)};

write(Protocol, Data) ->
    {error, Protocol, Data}.

File: script/make/protocol/lua/meta/DailyProtocol.lua
return {
    [12301] = {
        ["comment"] = "统计列表",
        ["write"] = {name = "data", type = "map", comment = "", explain = {

        }},
        ["read"] = {name = "data", type = "list", comment = "", explain = {
            {name = "count", type = "map", comment = "", explain = {
                {name = "type", type = "u32", comment = "统计类型", explain = {}},
                {name = "todayNumber", type = "u32", comment = "今日数量", explain = {}}
            }}
        }}
    },
    [12302] = {
        ["comment"] = "日常列表",
        ["write"] = {name = "data", type = "map", comment = "", explain = {

        }},
        ["read"] = {name = "data", type = "map", comment = "", explain = {
            {name = "daily", type = "map", comment = "", explain = {
                {name = "dailyId", type = "u32", comment = "日常ID", explain = {}},
                {name = "isAward", type = "u8", comment = "是否领取奖励", explain = {}}
            }},
            {name = "dailyActive", type = "map", comment = "", explain = {
                {name = "stageId", type = "u32", comment = "奖励阶段ID", explain = {}},
                {name = "score", type = "u32", comment = "活跃度", explain = {}}
            }}
        }}
    },
    [12303] = {
        ["comment"] = "领取日常奖励",
        ["write"] = {name = "data", type = "u32", comment = "日常ID", explain = {}},
        ["read"] = {name = "data", type = "rst", comment = "结果", explain = {}}
    },
    [12304] = {
        ["comment"] = "领取活跃度阶段奖励",
        ["write"] = {name = "data", type = "u32", comment = "阶段ID", explain = {}},
        ["read"] = {name = "data", type = "rst", comment = "结果", explain = {}}
    }
}